lib/std/heap.zig

```diff
@@ -38,37 +38,63 @@ const CAllocator = struct {
             pub const supports_malloc_size = true;
             pub const malloc_size = c.malloc_usable_size;
         }
+    else if (comptime @hasDecl(c, "_msize"))
+        struct {
+            pub const supports_malloc_size = true;
+            pub const malloc_size = c._msize;
+        }
     else
         struct {
             pub const supports_malloc_size = false;
         };
 
-    // The alignment guaranteed by malloc, the value matches the result of the C
-    // expression `alignof(max_alignment_t)`
-    const min_ptr_alignment = comptime std.math.max(
-        @alignOf(c_longdouble),
-        @alignOf(c_longlong),
-    );
+    pub const supports_posix_memalign = false and @hasDecl(c, "posix_memalign");
 
-    fn aligned_alloc(len: usize, alignment: usize) ?*c_void {
-        // The minimum alignment supported by both APIs is the size of a pointer
-        const eff_alignment = std.math.max(alignment, @sizeOf(usize));
+    fn get_header(ptr: [*]u8) *[*]u8 {
+        return @intToPtr(*[*]u8, @ptrToInt(ptr) - @sizeOf(usize));
+    }
 
-        if (builtin.os.tag == .windows) {
-            return c._aligned_malloc(len, eff_alignment);
+    fn aligned_alloc(len: usize, alignment: usize) ?[*]u8 {
+        if (supports_posix_memalign) {
+            // The minimum alignment supported posix_memalign is the pointer size
+            const eff_alignment = std.math.max(alignment, @sizeOf(usize));
+
+            var aligned_ptr: ?*c_void = undefined;
+            if (c.posix_memalign(&aligned_ptr, eff_alignment, len) != 0)
+                return null;
+
+            return @ptrCast([*]u8, aligned_ptr);
         }
 
-        var aligned_ptr: ?*c_void = undefined;
-        if (c.posix_memalign(&aligned_ptr, eff_alignment, len) != 0)
-            return null;
+        // Thin wrapper around regular malloc, overallocate to account for
+        // alignment padding and store the orignal malloc()'ed pointer before
```

Comment:
this potentially seems wasteful for large alignments; is `_aligned_malloc` really no better? (especially if you e.g. ask for a page aligned allocation)

Reply:
`_aligned_malloc` does exactly this, we can probably optimize this a bit later.